synapse/lib/iq.py

```diff
@@ -385,8 +397,63 @@ def getLoggerStream(self, logname):
         handler = logging.StreamHandler(stream)
         slogger = logging.getLogger(logname)
         slogger.addHandler(handler)
-        yield stream
-        slogger.removeHandler(handler)
+        try:
+            yield stream
+        except:  # pragma: no cover
+            raise
+        finally:
+            slogger.removeHandler(handler)
+
+    @contextlib.contextmanager
+    def setTstEnvars(self, **props):
+        '''
+        Set Environmental variables for the purposes of running a specific test.
+
+        Args:
+            **props: A kwarg list of envars to set. The values set are run
+            through str() to ensure we're setting strings.
+
+        Examples:
+            Run a test while a envar is set::
+
+                with self.setEnvars(magic='haha') as nop:
+                    ret = dostuff()
+                    self.true(ret)
+
+        Notes:
+            This helper explicitly sets and unsets values in os.environ, as
+            os.putenv does not automatically updates the os.environ object.
+
+        Yields:
+            None. This context manager yields None. Upon exiting, envars are
+            either removed from os.environ or reset to their previous values.
+        '''
+        old_data = {}
+        pop_data = set()
+        for key, valu in props.items():
+            v = str(valu)
+            oldv = os.environ.get(key, None)
+            if oldv:  # pragma: no cover
```

Comment:
Could you test this by doing a with block inside a with block for the same env var?